### ISSUE TITLE
suspendCancellableCoroutine not blocking in runBlocking(Test)

### ISSUE THREAD

**Issue by JustinTullgren** [Kotlin/kotlinx.coroutines#2152]
Kotlin Version: 1.3.61
Coroutines Version: 1.3.3
Environment: Android - Instrumentation Tests

Hi,

We have been migrating existing thread/callback based apis to use coroutines by way of `suspendCancellableCoroutine`.  We provide a suspend fun that returns the suspend coroutine which wraps our calls and resumes the continuation when the callback is invoked.  This all works well.

In testing however we are struggling to correctly and idiomatically block the test until the older thread/callback api is finished to perform testing on the result.  We use the TestCoroutineDispatcher, set Dispatchers.Main, and inject the test dispatcher throughout the code.

I believe I may be misunderstanding how suspendCancellableCoroutine and runBlockingTest work together and would appreciate any help.  We also are conscientiously not using backpressure in our channels until we completely migrate everything and we also hope to go to flows in the future if that has any impact on the scenario.

Scenario:
1. Setup Test Dispatcher and DI to use test dispatcher. 
2. Run app initialization code (coroutines are invoked as things get going) in a runBlocking statement before tests start.
3. Run a runBlockingTest to assert on a value that is computed via initialization code.

Expected: All coroutines ran in blocking/synchronous fashion and initialization value is computed before test.
Actual: suspendCancellableCoroutine does not block the test thread while waiting for the java thread callback to be invoked and the initialization value is not computed.

Here is a contrived sample showing the use case 

```
@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class Test {
    private val testCoroutineDispatcher = TestCoroutineDispatcher()

    private lateinit var sleeper: Sleeper

    @Before
    fun before() {
        Dispatchers.setMain(testCoroutineDispatcher)

        DispatchersProvider.io = testCoroutineDispatcher
        DispatchersProvider.main = testCoroutineDispatcher
        DispatchersProvider.default = testCoroutineDispatcher

        sleeper = Sleeper()

        runBlocking {
            // should set result to 1
            sleeper.updateResult(true)
        }
    }

    @Test
    fun test() = runBlockingTest {
        sleeper.result shouldBe 1
    }
}

suspend fun suspendSleep(input: Boolean): Int {
    return suspendCancellableCoroutine { continuation ->
        threadSleep(input) { error, result ->
            if (error != null) {
                continuation.resumeWithException(error)
            } else {
                continuation.resume(result)
            }
        }
    }
}

fun threadSleep(input: Boolean, callback: (Exception?, Int) -> Unit) {
    Thread {
        Thread.sleep(5000)
        val result = if (input) 1 else 0
        callback(null, result)
    }.start()
}

@ExperimentalCoroutinesApi
class Sleeper : CoroutineScope {
    var result: Int = 0
        private set

    private val job = SupervisorJob()
    override val coroutineContext: CoroutineContext
        get() = job + DispatchersProvider.main

    private val inputChannel = Channel<Boolean>(Channel.UNLIMITED)
    private val outputChannel = Channel<Int>(Channel.UNLIMITED)
    private val workerInput = BroadcastChannel<Boolean>(1)
    private val loop = launch(CoroutineName("Loop")) {
        withContext(DispatchersProvider.default) {
            val worker = createWorker(workerInput.openSubscription(), outputChannel)
            while (true) {
                select<Unit> {
                    inputChannel.onReceive {
                        workerInput.send(it)
                    }

                    outputChannel.onReceive {
                        result = it
                    }
                }
            }
        }
    }

    private fun createWorker(inputChannel: ReceiveChannel<Boolean>, outputChannel: SendChannel<Int>): Job =
        launch(CoroutineName("Worker") + DispatchersProvider.io) {
            for (input in inputChannel) {
                val result = suspendSleep(input)
                outputChannel.send(result)
            }
        }

    fun updateResult(newValue: Boolean) {
        inputChannel.offer(newValue)
    }
}

object DispatchersProvider {
    var io: CoroutineDispatcher = Dispatchers.IO
    var main: CoroutineDispatcher = Dispatchers.Main
    var default: CoroutineDispatcher = Dispatchers.Default
}

```

**Comment by elizarov** [comment]
Currently `runBlockingTest` does not intergrate with arbitrary sources of callbacks. I recommend using `runBlocking` to run tests of this kind. You can start with:

```
@Test
fun test() = runBlocking { 
    val dispatcher = coroutineContext[ContinuationInterceptor]
    // set your dispatcher provider
    // run your test
}
```

It will be looking nicer if you write your own `runTest` function that performs all the setup for your testing infrastructure and use it in your actual tests.

**Comment by JustinTullgren** [comment]
@elizarov Thank you for confirming.  I just have a couple of follow up questions based on your example.

`coroutineContext[ContinuationInterceptor]` returns a `ContinuationInterceptor` not a `CoroutineDispatcher`.  We inject `CoroutineDispatcher` everywhere, not the former. Most samples I have seen say to mock `CoroutineDispatcher`.  Is that incorrect?

Even if we do the above, do I still have to manually advance/update the interceptor?

**Comment by elizarov** [comment]
You can actually do `coroutineContext[CoroutineDispatcher]`. It returns the same object, just properly casted to `CoroutineDispatcher` type. I hope it helps.

**Comment by BenTilbrook** [comment]
It looks like `StateFlow` involves the use of `suspendCancellableCoroutine`. Does this mean `StateFlow` is not supported by `runBlockingTest`?

**Comment by elizarov** [comment]
@BenTilbrook As long as your code does not suspend at all or just uses `delay` to wait for a specific time, it will work with `runBlockingTest` just fine. The root of the problem in this issue is integration with 3rd party sources of callbacks that might resume the suspended coroutine. That is what `runBlockingTest` does not work with.

**Comment by BenTilbrook** [comment]
Thanks @elizarov. I suspected this was the case. I write a lot of tests using `runBlockingTest` and I sometimes see behavior involving `StateFlow` I can't explain, so it helps to have this clarified.

**Comment by qwwdfsad** [comment]
Fixed in 1.6.0-RC with `runTest`

**Comment by dkhalanskyjb** [comment]
Not really fixed in 1.6.0-RC. Trying to run this test will lead to the same error.

The issue is that `runTest` will still not wait for arbitrary coroutines that it doesn't know about. In this case, it won't know about `loop` and `worker`. `runTest` will only wait for the coroutines that were launched from its `TestScope`. In this case, however, launching `loop` and `worker` in the `TestScope` will just lead to the test hanging, as these coroutines are never completed.

Feel free to reopen if you have any questions.
